src/v/kafka/server/quota_manager.cc

```diff
@@ -510,100 +509,69 @@ ss::future<clock::duration> quota_manager::throttle_fetch_tp(
+quota_manager::get_global_map_for_testing() const {
+    return _global_map;
+}
+
 // erase inactive tracked quotas. windows are considered inactive if
 // they have not received any updates in ten window's worth of time.
 void quota_manager::gc() {
     vassert(
-      ss::this_shard_id() == _client_quotas.shard_id(),
+      ss::this_shard_id() == quotas_shard,
       "gc should only be performed on the owner shard");
     auto full_window = _default_num_windows() * _default_window_width();
     auto expire_threshold = clock::now() - 10 * full_window;
     ssx::background
       = ssx::spawn_with_gate_then(_gate, [this, expire_threshold]() {
-            return do_gc(expire_threshold);
+            return container()
+              .invoke_on_all([expire_threshold](quota_manager& qm) {
+                  return qm.do_local_gc(expire_threshold);
+              })
+              .then([this]() { return do_global_gc(); });
         }).handle_exception([](const std::exception_ptr& e) {
             vlog(klog.warn, "Error garbage collecting quotas - {}", e);
         });
 }
 
-namespace {
-
-/// A simplified copy paste of `std::set_intersection`. Copied here because we
-/// rely on the fact that we're allowed to have first1 == result for in-place
-/// set intersection in `do_gc()` which the contract of `std::set_intersection`
-/// does not allow despite that its implementation does.
-/// Requires: the inputs to be sorted
-/// Guarantees: the output to be sorted
-template<typename It1, typename It2, typename ItR>
-ItR set_intersection(It1 first1, It1 last1, It2 first2, It2 last2, ItR result) {
-    while (first1 != last1 && first2 != last2) {
-        if (*first1 < *first2) {
-            ++first1;
-        } else if (*first2 < *first1) {
-            ++first2;
-        } else { // *first1 == *first2
-            *result = *first1;
-            ++first1;
-            ++first2;
+ss::future<> quota_manager::do_global_gc() {
+    vassert(
+      ss::this_shard_id() == quotas_shard,
+      "do_global_gc() should only be called on the owner shard");
+
+    // Hold to lock to ensure there are no updates to the map while iterating
+    auto lock = co_await _global_map_mutex->get_units();
+
+    for (auto it = _global_map->begin(); it != _global_map->end();) {
+        auto& [key, value] = *it;
+        if (value.unique()) {
+            // The pointer in the global map is effectively a weak pointer in
+            // that we want to destroy the quota when only the global map holds
+            // a reference to it. The reason why we have a std::shared_ptr<>
+            // instead of a std::weak_ptr<> in the global map is to ensure that
+            // deallocation happens on shard 0 (here in do_global_gc).
+            vlog(client_quota_log.trace, "Global GC expiring key: {}", key);
+            it = _global_map->erase(it);
+        } else {
+            it++;
         }
+
+        co_await ss::maybe_yield();
```

Comment:
Prefer `ss::coroutine::maybe_yield` (see https://github.com/redpanda-data/redpanda/pull/16228#issue-2094493727 for background)

Reply:
Makes sense, thanks for the context! Done